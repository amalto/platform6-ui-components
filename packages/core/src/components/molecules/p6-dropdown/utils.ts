export function addDropdownClass(child: Element): Element {
  if (child.nodeName === 'HR') {
    child.classList.add('dropdown-divider');
    return child;
  }

  child.classList.add('dropdown-item');
  return child;
}
