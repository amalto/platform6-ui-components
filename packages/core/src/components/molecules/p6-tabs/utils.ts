export function getTabId(id: string): string {
  return `${id}-tab`;
}

export function isTabValid(tab: Element): tab is HTMLElement {
  return 'title' in tab && 'id' in tab;
}
