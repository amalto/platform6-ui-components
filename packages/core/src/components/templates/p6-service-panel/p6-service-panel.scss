@import '@creativebulma/bulma-tooltip/src/sass/index';

:host {
  display: block;

  div[slot='label'] {
    display: inline-flex;
    flex-direction: row;
    align-items: center;
    h3 {
      margin: 0;
      text-transform: uppercase;
    }
  }

  div[slot='header-actions'] {
    .service-update_state {
      display: inline-flex;
      flex-direction: row;
      align-items: center;
      gap: $margin-gap;

      .state {
        text-transform: uppercase;

        @each $name, $pair in $colors {
          $color: nth($pair, 1);
          $color-invert: nth($pair, 2);

          &.is-#{$name} {
            color: $color;

            &:hover,
            &.is-hovered {
              color: bulmaDarken($color, 2.5%);
            }
          }
        }
      }
    }

    p6-button ::slotted(::first-letter) {
      text-transform: uppercase;
    }
    ::slotted(p6-button::first-letter) {
      text-transform: uppercase;
    }

    p6-button::first-letter {
      text-transform: uppercase;
    }
  }
}
