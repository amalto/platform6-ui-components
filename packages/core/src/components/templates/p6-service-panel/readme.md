# p6-service-panel

<!-- Auto Generated Below -->

## Properties

| Property            | Attribute | Description            | Type                                                                                   | Default     |
| ------------------- | --------- | ---------------------- | -------------------------------------------------------------------------------------- | ----------- |
| `name` _(required)_ | `name`    | Name of the service    | `string`                                                                               | `undefined` |
| `status`            | `status`  | Status of the service  | `ServiceStatus.Restart \| ServiceStatus.Started \| ServiceStatus.Stopped \| undefined` | `undefined` |
| `version`           | `version` | Version of the service | `string \| undefined`                                                                  | `undefined` |

## Events

| Event              | Description                                      | Type                |
| ------------------ | ------------------------------------------------ | ------------------- |
| `p6RefreshService` | Fires when the user tries to refresh the service | `CustomEvent<void>` |
| `p6ReloadService`  | Fires when the user tries to reload the service  | `CustomEvent<void>` |
| `p6StartService`   | Fires when the user tries to start the service   | `CustomEvent<void>` |
| `p6StopService`    | Fires when the user tries to stop the service    | `CustomEvent<void>` |

## Dependencies

### Depends on

- [p6-panel](../../molecules/p6-panel)
- [p6-action](../../atoms/p6-action)
- [p6-icon](../../atoms/p6-icon)
- [p6-button](../../atoms/p6-button)

### Graph

```mermaid
graph TD;
  p6-service-panel --> p6-panel
  p6-service-panel --> p6-action
  p6-service-panel --> p6-icon
  p6-service-panel --> p6-button
  p6-panel --> p6-action
  p6-panel --> p6-icon
  style p6-service-panel fill:#f9f,stroke:#333,stroke-width:4px
```

---

_Built with [StencilJS](https://stenciljs.com/)_
