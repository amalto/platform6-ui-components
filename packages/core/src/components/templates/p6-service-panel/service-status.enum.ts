export enum ServiceStatus {
  Started = 'SERVICE_STATE_STARTED',
  Restart = 'SERVICE_STATE_STARTED_RESTART',
  Stopped = 'SERVICE_STATE_STOPPED',
}
