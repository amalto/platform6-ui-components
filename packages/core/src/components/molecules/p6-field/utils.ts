import { isP6Control, isP6NativeControl, P6Control, P6NativeControl } from '../../../shared/form/control';
import { isInDefaultSlot } from '../../../utils/component';

// eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/explicit-module-boundary-types
export function isControl(elmt: unknown): elmt is P6Control<unknown> | P6NativeControl {
  return isP6Control(elmt) || isP6NativeControl(elmt);
}

export function getControl(children: HTMLCollection): P6Control<unknown> | P6NativeControl | undefined {
  return Array.from(children).filter(isInDefaultSlot).find(isControl) as P6Control<unknown> | P6NativeControl | undefined;
}
