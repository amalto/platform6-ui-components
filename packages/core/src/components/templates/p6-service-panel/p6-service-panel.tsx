import { Component, ComponentInterface, Element, Event, EventEmitter, h, Host, Prop } from '@stencil/core';
import { Mode, Size } from '../../../shared/types';
import { getL10n, L10n } from '../../../utils/translations';
import { ServiceStatus } from './service-status.enum';

function actionHandler(emitter: EventEmitter | undefined): (event: Event) => void {
  return () => {
    emitter?.emit();
  };
}

@Component({
  tag: 'p6-service-panel',
  styleUrl: 'p6-service-panel.scss',
  shadow: true,
})
export class P6ServicePanel implements ComponentInterface {
  @Element() host!: HTMLP6ServicePanelElement;

  /**
   * Name of the service
   */
  @Prop() name!: string;

  /**
   * Version of the service
   */
  @Prop() version: string | undefined;

  /**
   * Status of the service
   */
  @Prop() status: ServiceStatus | undefined;

  /**
   * Fires when the user tries to reload the service
   */
  @Event({ eventName: 'p6ReloadService' }) reloadEmitter?: EventEmitter<void>;

  /**
   * Fires when the user tries to refresh the service
   */
  @Event({ eventName: 'p6RefreshService' }) refreshEmitter?: EventEmitter<void>;

  /**
   * Fires when the user tries to start the service
   */
  @Event({ eventName: 'p6StartService' }) startEmitter?: EventEmitter<void>;

  /**
   * Fires when the user tries to stop the service
   */
  @Event({ eventName: 'p6StopService' }) stopEmitter?: EventEmitter<void>;

  private l10n: L10n | undefined;

  async componentWillLoad(): Promise<void> {
    this.l10n = await getL10n(this.host);
  }

  render(): JSX.Element {
    const isStartServiceDisabled = ServiceStatus.Started === this.status || ServiceStatus.Restart === this.status;
    const isStopServiceDisabled = ServiceStatus.Stopped === this.status;
    // eslint-disable-next-line no-template-curly-in-string
    const titleTooltiptext = this.l10n?.serviceVersion.replace('${serviceVersion}', this.version || 'unknown');

    const statusTitle = this.l10n?.serviceStatus;
    const statusClass = { state: true, [`is-${this.statusColor}`]: true };

    return (
      <Host>
        <p6-panel hideable>
          <div slot="label">
            <h3 class="has-tooltip-arrow has-tooltip-bottom" data-tooltip={titleTooltiptext}>
              {this.name}
            </h3>
            <p6-action mode={Mode.primary} onClick={actionHandler(this.reloadEmitter)}>
              <p6-icon name="sync-alt" />
            </p6-action>
          </div>
          <div slot="header-actions">
            <div class="service-update_state">
              <p6-action onClick={actionHandler(this.refreshEmitter)}>
                <p6-icon name="sync-alt" />
              </p6-action>
              <span>{statusTitle}</span>
              {' : '}
              <span class={statusClass}>{this.statusLabel}</span>
            </div>
            <p6-button
              outlined
              size={Size.small}
              mode={isStartServiceDisabled ? Mode.default : Mode.primary}
              disabled={isStartServiceDisabled}
              onClick={actionHandler(this.startEmitter)}
            >
              <p6-icon name="play" />
              <span>{this.l10n?.start}</span>
            </p6-button>

            <p6-button
              outlined
              size={Size.small}
              mode={isStopServiceDisabled ? Mode.default : Mode.danger}
              disabled={isStopServiceDisabled}
              onClick={actionHandler(this.stopEmitter)}
            >
              <p6-icon name="stop" />
              <span>{this.l10n?.stop}</span>
            </p6-button>
          </div>
          <slot />
        </p6-panel>
      </Host>
    );
  }

  get statusLabel(): string {
    if (this.l10n === undefined || this.status === undefined) {
      return 'unknown';
    }

    switch (this.status) {
      case ServiceStatus.Started:
        return this.l10n.started;
      case ServiceStatus.Restart:
        return this.l10n.restart;
      case ServiceStatus.Stopped:
        return this.l10n.stopped;
      default:
        return 'unknown';
    }
  }

  get statusColor(): string {
    if (ServiceStatus.Started === this.status || ServiceStatus.Restart === this.status) {
      return 'success';
    }

    if (ServiceStatus.Stopped === this.status) {
      return 'danger';
    }

    return 'info';
  }
}
