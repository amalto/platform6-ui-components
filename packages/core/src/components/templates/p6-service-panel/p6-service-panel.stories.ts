import { ComponentProps, getElement, makeStory, Props } from '../../../shared/storybook';
import { P6ServicePanel } from './p6-service-panel';
import { ServiceStatus } from './service-status.enum';

const component = 'p6-service-panel';

export default {
  title: 'Templates/Service Panel',
  component,
};

const componentProps: ComponentProps = ['hideable', 'hidden'];

const getStoryField = (props?: Props<P6ServicePanel>): HTMLElement => {
  return getElement(component, [getElement('p6-empty', [])], props);
};

export const Default = makeStory<{
  name: string;
  version: string;
  status: ServiceStatus | undefined;
}>({
  componentProps,
  args: {
    name: 'service',
    version: '0.0.1',
    status: ServiceStatus.Started,
  },
  builder: args => getStoryField(args),
});
