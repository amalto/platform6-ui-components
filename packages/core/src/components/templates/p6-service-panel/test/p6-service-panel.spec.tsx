import { newSpecPage } from '@stencil/core/testing';
import { L10n } from '../../../../utils/translations';
import l10n_en from '../locales/p6-service-panel.i18n.en.json';
import l10n_fr from '../locales/p6-service-panel.i18n.fr.json';
import { P6ServicePanel } from '../p6-service-panel';

// This is the section where we mock `fetch`
const unmockedFetch = global.fetch;

function getDataFromUrl(url: string): L10n | null {
  if (url.startsWith('/locales/p6-service-panel.i18n.en.json')) {
    return l10n_en;
  }
  if (url.startsWith('/locales/p6-service-panel.i18n.fr.json')) {
    return l10n_fr;
  }
  return null;
}

beforeAll(() => {
  jest.spyOn(global, 'fetch').mockImplementation(url => {
    const l10n = getDataFromUrl(url as string);
    if (l10n === null) {
      return Promise.reject(new Error(`No locale file found for ${url}`));
    }
    return Promise.resolve(new Response(JSON.stringify(l10n)));
  });
});

afterAll(() => {
  global.fetch = unmockedFetch;
});

describe('p6-service-panel', () => {
  it('renders', async () => {
    const page = await newSpecPage({
      components: [P6ServicePanel],
      html: `<p6-service-panel></p6-service-panel>`,
    });
    expect(page.root).toMatchSnapshot();
  });
});
