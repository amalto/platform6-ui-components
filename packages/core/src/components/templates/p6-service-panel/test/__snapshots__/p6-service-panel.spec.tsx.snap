// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`p6-service-panel renders 1`] = `
<p6-service-panel>
  <mock:shadow-root>
    <p6-panel hideable="">
      <div slot="label">
        <h3 class="has-tooltip-arrow has-tooltip-bottom" data-tooltip="Service version: unknown"></h3>
        <p6-action mode="5">
          <p6-icon name="sync-alt"></p6-icon>
        </p6-action>
      </div>
      <div slot="header-actions">
        <div class="service-update_state">
          <p6-action>
            <p6-icon name="sync-alt"></p6-icon>
          </p6-action>
          <span>
            Service status
          </span>
          :
          <span class="is-info state">
            unknown
          </span>
        </div>
        <p6-button mode="5" outlined="" size="0">
          <p6-icon name="play"></p6-icon>
          <span>
            start
          </span>
        </p6-button>
        <p6-button mode="0" outlined="" size="0">
          <p6-icon name="stop"></p6-icon>
          <span>
            stop
          </span>
        </p6-button>
      </div>
      <slot></slot>
    </p6-panel>
  </mock:shadow-root>
</p6-service-panel>
`;
